v0.7.0 changelog and documentation (#853)
PennyLaneAI/catalyst: doc/changelog.md

<h3>New features</h3>

* Add support for accelerating classical processing via JAX with `catalyst.accelerate`.
  [(#805)](https://github.com/PennyLaneAI/catalyst/pull/805)

  Classical code that can be just-in-time compiled with JAX can now be seamlessly just
  in time compiled with `catalyst.accelerate` and included within QJIT-compiled functions.

>   `catalyst.accelerate` can be used as a
>   decorator without specifying a device:

Author: ```suggestion
```

  ```python

>   @accelerate(dev=jax.devices("gpu")[0])

Author: Huh, maybe this sentence can be deleted altogether, it makes no sense.

        return callback_fn(x), x

    @callback_fn.bwd
    def callback_fn_vjp(res, dy):
        # Accepts residuals from the forward pass, as well
        # as (one or more) cotangent vectors dy, and returns
        # a tuple of VJPs corresponding to each input parameter.

        def vjp(x, dy) -> (jax.ShapeDtypeStruct((2,), jnp.float64),):
            return (np.array([np.cos(x[0]) * dy * x[1], np.sin(x[0]) * dy]),)

        # The VJP function can also be a pure callback
        return catalyst.pure_callback(vjp)(res, dy)

    @qml.qjit
    @catalyst.grad
    def f(x):
        y = jnp.array([jnp.cos(x[0]), x[1]])
        return jnp.sin(callback_fn(y))
    ```

    ```pycon
    >>> x = jnp.array([0.1, 0.2])
    >>> f(x)
    array([-0.01071923,  0.82698717])
    ```

    Note that `@callback_fn.bwd` is compatible with the return type of `jax.vjp`,
    allowing for functions with `jax.vjp`-computable vector-Jacobian products
    to be easily supported:

    ```python
    @catalyst.pure_callback
    def expm(x) -> jax.ShapeDtypeStruct((2, 2), jnp.float64):
        return jax.scipy.linalg.expm(x)

    @expm.fwd
    def expm_fwd(x):

>         return jax.scipy.linalg.expm(x), x

Author: Yep, I think I agree with removing it

    @expm.bwd
    def expm_bwd(res, dy):

        @catalyst.accelerate
        def vjp(x, dy):
            _, vjp_func = jax.vjp(jax.scipy.linalg.expm, x)
            return vjp_func(dy)

        return vjp(res, dy)

    @qml.qjit
    @catalyst.jacobian
    def f(x):
        return expm(x)
    ```

    ```pycon
    >>> x = jnp.array([[0.1, 0.2], [0.3, 0.4]])
    >>> f(x)
    array([[[[1.12914705, 0.18555319],
         [0.12370212, 0.0129465 ]],
        [[0.12370212, 1.31470024],
         [0.008631  , 0.13664862]]],
       [[[0.18555319, 0.01941975],
         [1.31470024, 0.20497293]],
        [[0.0129465 , 0.20497293],
         [0.13664862, 1.51967317]]]])
    ```

* Catalyst now supports the 'dynamic one shot' method for simulating circuits with mid-circuit
  measurements, which compared to other methods, may be advantageous for circuits with many
  mid-circuit measurements executed for few shots.
  [(#5617)](https://github.com/PennyLaneAI/pennylane/pull/5617)
  [(#798)](https://github.com/PennyLaneAI/catalyst/pull/798)

  The dynamic one shot method evaluates dynamic circuits by executing them one shot at a time via

>   `catalyst.vmap`, sampling a dynamic execution path for each shot. This method only works for a

Author: Yeah I had the same thought as you. It was in the original changelog entry and I was on the fence

>   `catalyst.vmap`, sampling a dynamic execution path for each shot. This method only works for a

Author: I figure lets just keep it in for now, it's a nice reminder of vmap

  QNode executing with finite shots, and it requires the device to support mid-circuit measurements
  natively.

  This new mode can be specified by using the `mcm_method` argument of the QNode:

  ```python
  dev = qml.device("lightning.qubit", wires=5, shots=20)

  @qml.qjit(autograph=True)
  @qml.qnode(dev, mcm_method="one-shot")
  def circuit(x):

      for i in range(10):
          qml.RX(x, 0)
          m = catalyst.measure(0)

          if m:
              qml.RY(x ** 2, 1)

          x = jnp.sin(x)

      return qml.expval(qml.Z(1))
  ```

  Catalyst's existing method for simulating mid-circuit measurements remains
  available via `mcm_method="single-branch-statistics"`.

  When using `mcm_method="one-shot"`, the `postslect_mode` keyword argument can also
  be used, to specify whether postselection should continue until `shots`-number of
  postselected measurements are made (`"fill-shots"`), or whether execution should
  terminate once `shot` executions have occurred regardless of how many postselected

>   measurements have occurred (`"hw_like"`):

Author: Have updated this to

```markdown
  When using `mcm_method="one-shot"`, the `postselect_mode` keyword argument can also
  be used to specify whether the returned result should include `shots`-number of
  postselected measurements (`"fill-shots"`), or whether results should
  include all results, including invalid postselections (`"hw_like"`):
```

  ```

  Available devices can be retrieved via
  `jax.devices()`. If not provided, the default value of
  `jax.devices()[0]` as determined by JAX will be used.

* Catalyst callback functions, such as `pure_callback`, `debug.callback`, and `debug.print`, now
  all support autodifferentiation.
  [(#706)](https://github.com/PennyLaneAI/catalyst/pull/706)
  [(#782)](https://github.com/PennyLaneAI/catalyst/pull/782)
  [(#822)](https://github.com/PennyLaneAI/catalyst/pull/822)
  [(#834)](https://github.com/PennyLaneAI/catalyst/pull/834)

  - When using callbacks that do not return any values, such as `catalyst.debug.callback` and
    `catalyst.debug.print`, these functions are marked as 'inactive' and do not contribute to or
    affect the gradient of the function:

    ```python
    import logging
  
    log = logging.getLogger(__name__)
    log.setLevel(logging.INFO)
  
    @qml.qjit
    @catalyst.grad
    def f(x):
        y = jnp.cos(x)
        catalyst.debug.print("Debug print: y = {0:.4f}", y)
        catalyst.debug.callback(lambda _: log.info("Value of y = %s", _))(y)
        return y ** 2
    ```
  
    ```pycon
    >>> f(0.54)
    INFO:__main__:Value of y = 0.8577086813638242

>     Debug print: y = 0.8577

Reviewer: The order is not preserved? :thinking:

Author: Doesn't seem to be 🤔

    array(-0.88195781)
    ```

  - Callbacks that *do* return values and may affect the qjit-compiled functions

>     computation, such as `pure_callback`, must have custom gradients manually

Reviewer: ```suggestion
    computation, such as `pure_callback`, may have custom gradients manually
```

If the user uses a `pure_callback` somewhere that is not reachable from a `grad` op, then there is no need to register custom gradients. Not sure if you want this correction. Just thought it is an important distinction.

  
      for i in range(N):
          x = x + catalyst.disable_autograph(approximate_e)(10) / x ** i
  
      return x
  ```

  ```pycon
  >>> g(0.1, 10)
  array(4.02997319)
  ```

  Note that for Autograph to be disabled, the decorated function must be
  defined **outside** the qjit-compiled function. If it is defined within
  the qjit-compiled function, it will continue to be converted with Autograph.

  In addition, Autograph can also be disabled for all externally defined functions
  within a qjit-compiled function via the context manager syntax:

  ```python
  @qml.qjit(autograph=True)
  def g(x: float, N: int):
  
      for i in range(N):
          with catalyst.disable_autograph:
            x = x + approximate_e(10) / x ** i
  
      return x
  ```

* Support for including a list of (sub)modules to be whitelisted for autograph conversion.
  [(#725)](https://github.com/PennyLaneAI/catalyst/pull/725)

  Although library code is not meant to be targeted by Autograph conversion,
  it sometimes make sense to enable it for specific submodules that might
  benefit from such conversion:

  ```py
  @qjit(autograph=True, autograph_include=["excluded_module.submodule"])
  def f(x):

>     return excluded_module.submodule.func(x)

Author: I've added the following sentence: 

>   For example, this might be useful if importing functionality from PennyLane (such as
  a transform or decomposition), and would like to have Autograph capture and convert
  associated control flow.

>   shot in a shots-based execution mode.

Reviewer 2: I think this wording is outdated, now that we do have the one-shot mode it is possible to produce an entire mcm array with it. 

Also might be worth considering as a new feature?

Author: https://github.com/PennyLaneAI/catalyst/pull/853/commits/3dc88f9fe0a77fef7bd1f20cb880f4fa20766261

Reviewer 2: Nice addition 👍

  [(#731)](https://github.com/PennyLaneAI/catalyst/pull/731)

* The Catalyst frontend now supports Python logging through PennyLane's `qml.logging` module.

>   [(#660)](https://github.com/PennyLaneAI/catalyst/pull/660)

Reviewer 2: Is this feature documented somewhere that we can link to?

Author: I couldn't find anything 🤔 @mlxd?

Reviewer 2: Looks like we do: https://docs.pennylane.ai/en/stable/introduction/logging.html

Author: cbacdec3a0dfe23bd18251c66625f68c8184d063

* Catalyst now performs a stricter validation of the wire requirements for devices. In particular,
  only integer, continuous wire labels starting at 0 are allowed.
  [(#784)](https://github.com/PennyLaneAI/catalyst/pull/784)

* Catalyst no longer disallows quantum circuits with 0 qubits.
  [(#784)](https://github.com/PennyLaneAI/catalyst/pull/784)

> * Added support for IsingZZ gate in Catalyst frontend. Previously, the IsingZZ gate would be
>   decomposed into a CNOT and RZ gates. However, this is not needed as the PennyLane-Lightning
>   simulator supports this gate.

Reviewer 2: ```suggestion
* Added support for `IsingZZ` as a native gate in Catalyst. Previously, the IsingZZ gate would be
  decomposed into a CNOT and RZ gates, even if a device supported it.
```

  [(#730)](https://github.com/PennyLaneAI/catalyst/pull/730)

* Functions that have been annotated with return type

>   annotations will now correctly compile with `@qjit`.

Reviewer 2: Is this a bug fix?

Author: Good catch!

  [(#751)](https://github.com/PennyLaneAI/catalyst/pull/751)

* The `vmap`,`qjit`, `mitigate_with_zne` and gradient decorators now follow
  a unified pattern that uses a callable class implementing the decorator's logic. This
  prevents having to excessively define functions in a nested fashion, and allows them

>   to be used as decorators.

Reviewer 2: Tempted to say this is an internal change, at least the first part (excessively define functions in a nested fashion), the second part about enabling the same usage pattern across all decorators is user-facing, but the entry doesn't describe what that pattern is.

Essentially, it boils down to being able to supply keyword arguments to the decorator without the need for partial, which is quite convenient. So all our decorators allow both of:
- `@decorator`
- `@decorator(param1 = ...)`

Author: done!

  [(#758)](https://github.com/PennyLaneAI/catalyst/pull/758)
  [(#761)](https://github.com/PennyLaneAI/catalyst/pull/761)

>   [(#762)](https://github.com/PennyLaneAI/catalyst/pull/762)

Reviewer 2: The next entry (PR 712) is internal for sure. Considering it is about tests I would even remove it entirely.

Reviewer 2: Typo for 772:
```
* The built-in instrumentation with `detailed` output will no longer report the cumulative time for
  MLIR pipelines, since the cumulative time was being reported as just another step alongside
  individual timings for each pipeline.
```

Reviewer 2: 790 is also an internal change. I would expand it a bit to clarify context:
```
* Measurement process primitives for Catalyst's JAXPR representation now have a standardized
  call signature so that `shots` and `shape` can both be provided as keyword arguments.
```

Reviewer 2: 789 is a bit vague:
```
* The finite difference method for differentiation is now always allowed, even on functions with
  mid-circuit measurements, callbacks without custom derivates, or other operations that cannot
  be differentiated via traditional autodiff.
```

Reviewer 2: 767 should be highlighted further up I think, being that we officially support a new platform now with this release :)  The part about the github workflow is irrelevant.

Author: - I've removed 712 entirely
- 772 fixed
- 790 I've moved to internal changes and updated
- 789 updated
- 767 I've actually updated and moved to the new features section:

  * Catalyst is now officially support on Linux aarch64, with pre-built binaries
  available on PyPI; simply `pip install pennylane-catalyst` on Linux aarch64 systems.
  [(#767)](https://github.com/PennyLaneAI/catalyst/pull/767)

* Binary distributions for Linux are now based on `manylinux_2_28` instead of `manylinux_2014`.
  As a result, Catalyst will only be compatible on systems with `glibc` versions `2.28` and above
  (e.g., Ubuntu 20.04 and above).
  [(#663)](https://github.com/PennyLaneAI/catalyst/pull/663)

> <h3>Bug fixes</h3>

Reviewer 2: 801 needs information that this is about the runtime:
```
* An issue in the Lightning backend for the Catalyst runtime has been fixed that would only compute
  approximate probabilities when implementing mid-circuit measurements. As a result, low shot numbers 
  would lead to unexpected behaviours or projections on zero probability states.
  Probabilities for mid-circuit measurements are now always computed analytically.
```

Author: thanks, this is much more informative!

  register.
  [(#784)](https://github.com/PennyLaneAI/catalyst/pull/784)

* `jax.scipy.linalg.expm` is now supported within qjit-compiled functions.
  [(#733)](https://github.com/PennyLaneAI/catalyst/pull/733)
  [(#752)](https://github.com/PennyLaneAI/catalyst/pull/752)

  This required correctly linking openblas routines necessary for `jax.scipy.linalg.expm`.
  In this bug fix, four openblas routines were newly linked and are now discoverable by
  `stablehlo.custom_call@<blas_routine>`. They are `blas_dtrsm`, `blas_ztrsm`, `lapack_dgetrf`,
  `lapack_zgetrf`.

> * Correctly recording types of constant array when lowering `catalyst.grad` to mlir.

Author: ```suggestion
* Fixes a bug where QNodes that contained `QubitUnitary` with a complex matrix
  would error during gradient computation.
```

  [(#778)](https://github.com/PennyLaneAI/catalyst/pull/778)

* Callbacks can now return types which can be flattened and unflattened.
  [(#812)](https://github.com/PennyLaneAI/catalyst/pull/812)

* `catalyst.qjit` and `catalyst.grad` now get `__name__` from `functools.partial`.

>   [(#820)](https://github.com/PennyLaneAI/catalyst/pull/820)

Author: @tzunghanjuang could you describe what this change was for?

Reviewer 3: It's to make catalyst able to get function's name. For cases that calculate `grad` from functions wrapped by partial trigger no attribute `__name__` error.
```python
        def fn(x, y):
            return x * y

        partial_fn = partial(fn, y=1)

        @qjit
        def grad_partial_fn(x):
            return grad(partial_fn)(x)
```

Author: ```suggestion
* `catalyst.qjit` and `catalyst.grad` now work correctly with
  functions that have been wrapped with `functools.partial`.
  [(#820)](https://github.com/PennyLaneAI/catalyst/pull/820)
```

>   register.

Reviewer 2: Line just above: "if *a* qubit"

  [(#784)](https://github.com/PennyLaneAI/catalyst/pull/784)

* `jax.scipy.linalg.expm` is now supported within qjit-compiled functions.
  [(#733)](https://github.com/PennyLaneAI/catalyst/pull/733)
  [(#752)](https://github.com/PennyLaneAI/catalyst/pull/752)

  This required correctly linking openblas routines necessary for `jax.scipy.linalg.expm`.
  In this bug fix, four openblas routines were newly linked and are now discoverable by
  `stablehlo.custom_call@<blas_routine>`. They are `blas_dtrsm`, `blas_ztrsm`, `lapack_dgetrf`,
  `lapack_zgetrf`.

* Fixes a bug where QNodes that contained `QubitUnitary` with a complex matrix
  would error during gradient computation.
  [(#778)](https://github.com/PennyLaneAI/catalyst/pull/778)

* Callbacks can now return types which can be flattened and unflattened.
  [(#812)](https://github.com/PennyLaneAI/catalyst/pull/812)

> * `catalyst.qjit` and `catalyst.grad` now work correctly with

Reviewer 2: ```suggestion
* `catalyst.qjit` and `catalyst.grad` now work correctly on
```

  functions that have been wrapped with `functools.partial`.
  [(#820)](https://github.com/PennyLaneAI/catalyst/pull/820)

> <h3>Internal changes</h3>

Reviewer 2: 771 and 768 need to be updated. Instead of generating a new class `Controlled`/`Adjoint` that inherits from both catalyst and pennylane classes, using adjoint/ctrl on a single operation will now directly dispatch to the pennylane class.

Reviewer 2: 739 should have a different description or be removed entirely. I don't think that linting needs to appear in the change log.

Reviewer 2: We might want to consider moving 711 up out of the internal changes. It is somewhat useful to know that the operation object can now be accessed from the control flow decorators, especially for developers writing transforms.

Reviewer 2: Program verification 626/783 might even be considered a new feature no? Definitely much too user-facing for an internal change. It also deserves a better description than just one line.

Author: - Have updated 771 and 768, and also added in a link to 802. Note that 771/768/802 are also listed in the Improvements section, so we could potentially remove their listings in Internal Changes.

- 739 has been removed

- I've moved 711 to improvements. Something that comes to mind -- is this documented anywhere outside of the changelog? It might get lost from awareness otherwise

- 626/783: I've moved this to the top of improvements, and added a code example

Reviewer 2: >I've moved 711 to improvements. Something that comes to mind -- is this documented anywhere outside of the changelog? It might get lost from awareness otherwise

Good point. It should probably be included in the for_loop/while_loop/cond docstring if it isn't already!
